builtins/amp-img.js

```diff
@@ -178,13 +178,24 @@ export class AmpImg extends BaseElement {
     // It is important to call this before setting `srcset` attribute.
     this.maybeGenerateSizes_(/* sync setAttribute */ true);
     this.propagateAttributes(ATTRIBUTES_TO_PROPAGATE, this.img_);
+    this.propagateDataset_();
     guaranteeSrcForSrcsetUnsupportedBrowsers(this.img_);
     this.applyFillContent(this.img_, true);
     propagateObjectFitStyles(this.element, this.img_);
 
     this.element.appendChild(this.img_);
   }
 
+  /**
+   * Propagate the dataset key/value pairs in the `amp-img` element to the
+   * backing `img` element.
+   */
+  propagateDataset_() {
+    for (const key in this.element.dataset) {
```

Comment:
This can get out of sync very easily if data attribute is removed from the amp-img.

Reply:
You just have to listen in `mutatedAttributesCallback`.

Reply:
done. do you mind taking another look? thanks. i just delete and reapply everything again. i wasn't sure if there was anything more granular like a delete entry on the object that is passed into `mutatedAttributesCallback`